modules/distributor/usage/tracker.go

```diff
@@ -0,0 +1,381 @@
+	values     []string // Used to capture values and guaranteed to be a matching length.
+}
+
+// GetBuffersForDimensions takes advantage of the fact that the configuration for a tracker
+// changes slowly.  Reuses buffers from the previous call when the dimensions are the same.
+func (t *tenantUsage) GetBuffersForDimensions(dimensions []string) ([]string, []string) {
+	reset := false
+
+	if len(dimensions) == len(t.dimensions) {
+		// Any change to dimensions?
+		for i := range dimensions {
+			if dimensions[i] != t.dimensions[i] {
+				reset = true
+				break
+			}
+		}
+	} else {
+		reset = true
+	}
+
+	if reset {
+		t.dimensions = dimensions
+		t.labels = make([]string, len(dimensions))
+		t.values = make([]string, len(dimensions))
+		for i, d := range dimensions {
+			t.labels[i] = strutil.SanitizeFullLabelName(d)
+		}
+	}
+
+	return t.labels, t.values
+}
+
+func (t *tenantUsage) getSeries(labels, values []string, maxCardinality uint64) *bucket {
+	h := hash(labels, values)
+
+	b := t.series[h]
+	if b == nil {
+		// Before creating a new series, check for cardinality limit.
+		if uint64(len(t.series)) >= maxCardinality {
+			// Overflow
+			// It goes into the unlabeled bucket
+			// TODO - Do we want to do something else?
+			clear(values)
+			h = emptyHash
+			b = t.series[h]
+		}
+	}
+
+	if b == nil {
+		// First encounter with this series. Initialize it.
+		l, v := nonEmpties(labels, values)
+		b = &bucket{
+			// Metric description - constant for this pass now that the dimensions are known
+			descr:  prometheus.NewDesc("tempo_usage_tracker_bytes_received_total", "bytes total received with these attributes", l, t.constLabels),
+			labels: v,
+		}
+		t.series[h] = b
+	}
+	return b
+}
+
+type Tracker struct {
+	mtx      sync.Mutex
+	name     string
+	tenants  map[string]*tenantUsage
+	labelsFn tenantLabelsFunc
+	maxFn    tenantMaxFunc
+	reg      *prometheus.Registry
+	cfg      Config
+}
+
+func NewTracker(cfg Config, name string, labelsFn tenantLabelsFunc, maxFn tenantMaxFunc) (*Tracker, error) {
+	u := &Tracker{
+		cfg:      cfg,
+		name:     name,
+		tenants:  make(map[string]*tenantUsage),
+		labelsFn: labelsFn,
+		maxFn:    maxFn,
+		reg:      prometheus.NewRegistry(),
+	}
+
+	err := u.reg.Register(u)
+	if err != nil {
+		return nil, err
+	}
+
+	go u.PurgeRoutine()
+
+	return u, nil
+}
+
+// getTenant must be called under lock.
+func (u *Tracker) getTenant(tenant string) *tenantUsage {
+	data := u.tenants[tenant]
+	if data == nil {
+		data = &tenantUsage{
+			series: make(map[uint64]*bucket),
+			constLabels: prometheus.Labels{
+				tenantLabel:  tenant,
+				trackerLabel: u.name,
+			},
+		}
+		u.tenants[tenant] = data
+	}
+	return data
+}
+
+func (u *Tracker) Observe(tenant string, batches []*v1.ResourceSpans) {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	dimensions := u.labelsFn(tenant)
+	if len(dimensions) == 0 {
+		// Not configured
+		// TODO - Should we put it all in the unattributed bucket instead?
```

Comment:
I would lean on the side of defaulting to service name, that seems more useful then the unattributed bucket 🤔

Reply:
Whatever is decided here, it should be the same for all our DBs.

Reply:
I think for now we won't have any default behavior. It is still possible to configure default behavior yourself, by setting the default-tenant overrides (tenantID=`*`).  The tracker is also disabled by default since it is only relevant to large multi-tenant scenarios.

Reply:
I think the the default behavior of nothing is the right choice. In a large multi-tenant cell we may only want this turned on for a few tenants.

Should we move this outside lock since it's called on every push and the normal case will likely be to return here?

Reply:
Moved